src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,455 @@
+namespace ts.codefix {
+    const fixId = "convertToAsyncFunction";
+    const errorCodes = [Diagnostics.This_may_be_converted_to_an_async_function.code];
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context: CodeFixContext) {
+            const changes = textChanges.ChangeTracker.with(context, (t) => convertToAsyncFunction(t, context.sourceFile, context.span.start, context.program.getTypeChecker(), context));
+            return [createCodeFixAction(fixId, changes, Diagnostics.Convert_to_async_function, fixId, Diagnostics.Convert_all_to_async_functions)];
+        },
+        fixIds: [fixId],
+        getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, err) => convertToAsyncFunction(changes, err.file, err.start, context.program.getTypeChecker(), context)),
+    });
+    function convertToAsyncFunction(changes: textChanges.ChangeTracker, sourceFile: SourceFile, position: number, checker: TypeChecker, context: CodeFixContextBase): void {
+        // get the function declaration - returns a promise
+        const functionToConvert: FunctionLikeDeclaration = getContainingFunction(getTokenAtPosition(sourceFile, position)) as FunctionLikeDeclaration;
+        if (!functionToConvert) {
+            return;
+        }
+
+        // add the async keyword
+        changes.insertModifierBefore(sourceFile, SyntaxKind.AsyncKeyword, functionToConvert);
+
+        const synthNamesMap: Map<[Identifier, number]> = createMap(); // number indicates the number of times it is used after declaration
+        const lastDotThenMap: Map<boolean> = createMap();
+
+        const functionToConvertRenamed: FunctionLikeDeclaration = renameCollidingVarNames(functionToConvert, checker, synthNamesMap, context);
+        findLastDotThens(functionToConvertRenamed, lastDotThenMap, checker);
+
+        const returnStatements = getReturnStatementsWithPromiseCallbacks(functionToConvertRenamed);
+        let allNewNodes: Map<Node[]> = createMap();
+        for (const statement of returnStatements) {
+            if (isCallExpression(statement)) {
+                const newNodes = parseCallback(statement, checker, statement, synthNamesMap, lastDotThenMap, context);
+                if (newNodes.length) {
+                    allNewNodes = allNewNodes.set(String(getNodeId(statement)), newNodes);
+                }
+            }
+            else {
+                forEachChild(statement, function visit(node: Node) {
+                    if (isCallExpression(node)) {

```

Comment:
This block seems to be same modulo names as the block above - refactor out into a shared function?